### PR TITLE
Procedures update for optional features

### PR DESCRIPTION
**Pull request by Wind4Greg**
This PR addresses issues:   https://github.com/w3c/vc-di-bbs/issues/153,  https://github.com/w3c/vc-di-bbs/issues/156 and also https://github.com/w3c/vc-di-bbs/issues/150 and https://github.com/w3c/vc-di-bbs/issues/157.

It does this by clarifying which optional feature, if any, is being used via an `featureOption` variable, feature specific base proof and derived proof header bytes, and extending procedure/function descriptions to clarify inputs/outputs. In addition *informative* summary tables have been added to the section describing the optional features. 

Note that some optional features are dependent on IETF APIs that are yet to be frozen. These have been notated.

***
<a href="https://pr-preview.s3.amazonaws.com/Wind4Greg/vc-di-bbs/pull/158.html" title="Last updated on Apr 28, 2024, 4:09 PM UTC (bae66c2)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-di-bbs/158/49c52b5...Wind4Greg:bae66c2.html" title="Last updated on Apr 28, 2024, 4:09 PM UTC (bae66c2)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1290,10 +1499,13 @@ <h4>Base Proof Serialization (bbs-2023)</h4>
 in the Data Integrity [[VC-DATA-INTEGRITY]] specification,
 <a data-cite="vc-data-integrity#algorithms">
 Section 4: Algorithms</a>. Required inputs are
-cryptographic hash data (|hashData|) and
-<em>proof options</em> (|options|).
-Optional inputs include a |commitment_with_proof| byte array and/or a
-|use_pseudonyms| boolean.
+cryptographic hash data (|hashData|),
+<em>proof options</em> (|options|), |featureOption|, and if required
```
Comment (TallTed) [inline index.html:1503]:
```suggestion
<em>proof options</em> (|options|), |featureOption|, and, if required,
```

File: index.html [hunk 2]
```diff
@@ -1454,41 +1688,49 @@ <h4>Verify Derived Proof (bbs-2023)</h4>
...
+obtained from the UTF-8 encoding of the elements of the |nonMandatory| array.
             </li>
             <li>
 Initialize |verified| to the result of applying the verification
-algorithm below, depending on whether the |pseudonym| value is empty.
+algorithm below, depending the |featureOption| value.
             <ol class="algorithm">
               <li>
-If the |pseudonym| value is empty, initialize |verified| to the result of
+If the |featureOption| equals `"baseline"`, Initialize |verified| to the result of
```
Comment (TallTed) [inline index.html:1708]:
```suggestion
If the |featureOption| equals `"baseline"`, initialize |verified| to the result of
```

File: index.html [hunk 3]
```diff
@@ -1648,7 +1891,192 @@ <h3>Pseudonyms with Hidden PID</h3>
...
+      </section>
+      <section class="informative">
+        <h3>Optional Feature Summary</h3>
+        <p>
+This section provides a summaries of the inputs, outputs, proof serialiation,
+tasks, and procedures for "baseline" BBS proofs as well as those for the
+optional features. By <em>baseline</em> BBS we mean BBS base and derived proofs
+without additional features. All the optional features are "additive" in the
+sense that some additional input, task, or, output is generated in
+<em>addition</em> to that of the "baseline" BBS signatures/proofs.
```
Comment (TallTed) [inline index.html:1903]:
```suggestion
This section provides summaries of the inputs, outputs, proof serialiation,
tasks, and procedures for "baseline" BBS proofs as well as those for the
optional features. By <em>baseline</em> BBS, we mean BBS base and derived proofs
without additional features. All the optional features are "additive" in the
sense that some additional input, task, or output is generated
<em>in addition</em> to those of the "baseline" BBS signatures/proofs.
```

**Review by TallTed — CHANGES_REQUESTED** [review]

File: index.html [hunk 4]
```diff
@@ -627,9 +679,10 @@ <h4>parseBaseProofValue</h4>
 The following algorithm parses the components of a `bbs-2023` selective
 disclosure base proof value. The required input is a proof value
 (|proofValue|). A single object, <em>parsed base proof</em>, containing
-five  or seven elements, using the names "bbsSignature", "bbsHeader",
+six or seven elements, using the names "bbsSignature", "bbsHeader",
 "publicKey",
-"hmacKey", "mandatoryPointers", and optional feature parameters "pid" and
+"hmacKey", "mandatoryPointers", "featureOption", and possibly optional feature
+parameters "pid" and
 "signer_blind" is produced  as output.
```
[the remark this answers is not in the document]
Comment (Wind4Greg) [reply index.html:686]: The tricky thing here is that depending on the "feature option" you will get a *signer_blind* (anonymous holder binding and pseudonym hidden pid cases), or the *pid* (pseudonym issuer-known pid case). I summarized this in Table 2 in the "Optional Feature Summary" of this PR but didn't know the best way to say it.

File: index.html [hunk 5]
```diff
@@ -674,50 +773,49 @@ <h4>createDisclosureData</h4>
...
           </p>
 
           <ol class="algorithm">
             <li>
-Initialize `bbsSignature`, `bbsHeader`, `publicKey`, `hmacKey`,
-`mandatoryPointers`, and the optional feature parameters `pid` and
-`signer_blind` to the values of the associated properties in the object
+Initialize |bbsSignature|, |bbsHeader|, |publicKey|, |hmacKey|,
+|mandatoryPointers|, and the optional feature parameters |pid| and
+|signer_blind| to the values of the associated properties in the object
```
Comment (TallTed) [inline index.html:789]: Does it matter that `|pid|` and `|signer_blind|` are optional, here? Or should they just always be initialized as described?
```suggestion
Initialize |bbsSignature|, |bbsHeader|, |publicKey|, |hmacKey|,
|mandatoryPointers|, |pid|, and
|signer_blind| to the values of the associated properties in the object
```

File: index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (TallTed) [reply index.html:686]: OK, so you will get either _`signer_blind`_ or _`pid`_ or neither, but never both?
```suggestion
parameter "pid" or
"signer_blind", is produced  as output.
```

File: index.html (hunk 5 above)
Comment (Wind4Greg) [reply index.html:789]: Hi @TallTed, in section "4.4 Optional Feature Summary" Table 2, I summarize that under different options the `pid` or (exclusive) the `signer_blind`, or neither would be used.  That is `paseBaseProof` returns different items under different options (this is called in step 1 of `createDisclosureData`. These are then used later in step 10 of  `createDisclosureData` where different processing is done based on the `featureOption` value.

File: index.html [hunk 6]
```diff
@@ -590,20 +590,72 @@ <h4>serializeBaseProofValue</h4>
...
 
           <ol class="algorithm">
             <li>
-Initialize a byte array, `proofValue`, that starts with the BBS base proof
-header bytes `0xd9`, `0x5d`, and `0x02`.
+Depending upon the value of the |featureOption|, set up the |proofValue| as
+follows.
             </li>
             <li>
+If |featureOption| equals `"baseline"`:
```
Comment (dlongley) [inline index.html:606]: I think we might want to shorten this to just `feature` (or another single word name like "mode" or something that we come to consensus on), but I wouldn't hold up this PR for it. I think this PR is an approvement and am in support.

File: index.html [hunk 7]
```diff
@@ -914,14 +1036,66 @@ <h4>serializeDerivedProofValue</h4>
...
+                    </li>
+                    <li>
+Initialize |components| to an array with elements containing the values of
+|bbsProof|, |compressedLabelMap|, |mandatoryIndexes|, |selectiveIndexes|, and
+|presentationHeader|.
+                    </li>
+                  </ol>
+                </li>
+                <li>
+If |featureOption| equals `"anonymous_holder_binding"`:
```
Comment (dlongley) [inline index.html:1056]: Using snake_case here is inconsistent with our other names, so I'd prefer it be camel case or hyphenated (or whatever else matches our current style elsewhere), but again wouldn't hold up the PR for that.

**Review by dlongley — APPROVED** [review]

**Review by msporny — APPROVED** [review]

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

File: index.html (hunk 6 above)
Comment (Wind4Greg) [reply index.html:606]: Hi @dlongley, agree. Picking good names is hard. Currently have a lot of names with *option* in it. Simpler the better.

File: index.html (hunk 7 above)
Comment (Wind4Greg) [reply index.html:1056]: Hi @dlongley, this is a bit of a mess and extends to variable names, and the IETF specs too (mixing camel and snake case). I would be nice to choose *camel* or *snake* case (or whatever) throughout or have a established naming convention for functions, variable, constants,  constant strings, etc. I'm more than happy to clean up everything according to a nice convention. Does W3C have anything written? I'm not strongly optioned on what the conventions should be.
